EDDiscovery/UserControls/UserControlScanGrid.cs

```diff
@@ -0,0 +1,206 @@
+        }
+
+        public override void Init()
+        {
+            uctg.OnTravelSelectionChanged += Display;
+            discoveryform.OnNewEntry += NewEntry;
+        }
+
+        public override void ChangeCursorType(IHistoryCursor thc)
+        {
+            uctg.OnTravelSelectionChanged -= Display;
+            uctg = thc;
+            uctg.OnTravelSelectionChanged += Display;
+        }
+
+        public override void Closing()
+        {
+            uctg.OnTravelSelectionChanged -= Display;
+            discoveryform.OnNewEntry -= NewEntry;
+        }
+
+        public override void InitialDisplay()
+        {
+            Display(uctg.GetCurrentHistoryEntry, discoveryform.history);
+        }
+
+        public void NewEntry(HistoryEntry he, HistoryList hl)               // called when a new entry is made.. check to see if its a scan update
+        {
+            // if he valid, and last is null, or not he, or we have a new scan
+            if (he != null && (last_he == null || he != last_he || he.EntryType == JournalTypeEnum.Scan))
+            {
+                last_he = he;
+                DrawSystem();
+            }
+        }
+
+        private void Display(HistoryEntry he, HistoryList hl)            // Called at first start or hooked to change cursor
+        {
+            if (he != null && (last_he == null || he.System != last_he.System))
+            {
+                last_he = he;
+                DrawSystem();
+            }
+        }
+
+        void DrawSystem()   // draw last_sn, last_he
+        {
+            dataGridViewScangrid.Rows.Clear();
+
+            if (last_he == null)
+            {
+                SetControlText("No Scan");
+                return;
+            }
+
+
+            StarScan.SystemNode last_sn = discoveryform.history.starscan.FindSystem(last_he.System, true);
+
+            SetControlText((last_sn == null) ? "No Scan" : ("Brief Scan Summary for " + last_sn.system.name));
+
+            if (last_sn != null)
+            {
+                List<StarScan.ScanNode> all_nodes = new List<StarScan.ScanNode>();
+                foreach (StarScan.ScanNode starnode in last_sn.starnodes.Values)
+                {
+                    all_nodes = Flatten(starnode, all_nodes);
+                }
+
+
+                // flatten tree of scan nodes to prepare for listing
+                foreach (StarScan.ScanNode sn in all_nodes)
+                {
+
+                    
+                    // create the grid data
+
+                    // populate the body class
+                    StringBuilder bdClass = new StringBuilder();
+                    
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.PlanetClass != null)
+                    {
+                        bdClass.Append(sn.ScanData.PlanetClass);
+                    }
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.StarTypeText != null)
```

Comment:
It's a good idea to group similar conditionals together. It's both shorter and clearer, as well as (insignificantly) faster:
```CS
if (sn.ScanData != null && sn.ScanData.BodyName != null)
{
    if (sn.ScanData.PlanetClass != null)
        bdClass.Append(sn.ScanData.PlanetClass);
    if (sn.ScanData.StarTypeText != null)
        bdClass.Append(sn.ScanData.StarTypeText);
}
```

Reply:
Yes, I was thinking in a way to have a more cleaner and elegant code. I'll rewrite it following your suggestion.

Reply:
Done.